pkg/compute/bidder.go

```diff
@@ -165,37 +221,141 @@ func (b Bidder) ReturnBidResult(
+		reason:              resourceResponse.reason,
+		calculatedResources: resourceResponse.calculatedResources,
+	}, nil
+}
+
+func (b Bidder) runSemanticBidding(
+	ctx context.Context,
+	job *models.Job,
+) (*bidStrategyResponse, error) {
+	// ask the bidding strategy if we should bid on this job
+	request := bidstrategy.BidStrategyRequest{
+		NodeID:   b.nodeID,
+		Job:      *job,
+		Callback: b.getApproveURL(),
+	}
+
+	// assume we are bidding unless a request is rejected
+	shouldBid := true
+	// assume we're not waiting to bid unless a request indicates so.
+	shouldWait := false
+	reasons := make([]string, 0, len(b.semanticStrategy))
+	for _, s := range b.semanticStrategy {
+		// TODO(forrest): this can be parallelized with a wait group, although semantic checks ought to be quick.
+		strategyType := reflect.TypeOf(s).String()
+		resp, err := s.ShouldBid(ctx, request)
+		if err != nil {
+			errMsg := fmt.Sprintf("bid strategy: %s failed", strategyType)
+			log.Ctx(ctx).Error().Err(err).Msgf(errMsg)
+			// NB: failure here results in a callback to OnComputeFailure
+			return nil, errors.Wrap(err, errMsg)
+		}
+		log.Ctx(ctx).Info().
+			Str("Job", request.Job.ID).
+			Str("strategy", strategyType).
+			Bool("bid", resp.ShouldBid).
+			Bool("wait", resp.ShouldWait).
+			Str("reason", resp.Reason).
+			Msgf("bit strategy response")
+
+		if resp.ShouldWait {
+			shouldWait = true
+			reasons = append(reasons, fmt.Sprintf("waiting to bid: %s", resp.Reason))
+		} else if !resp.ShouldBid {
+			shouldBid = false
+			reasons = append(reasons, fmt.Sprintf("rejected bid: %s", resp.Reason))
+		}
 	}
-	resourceResponse, err := b.resourceStrategy.ShouldBidBasedOnUsage(ctx, request, *resourceUsage)
+
+	return &bidStrategyResponse{
+		bid:    shouldBid,
+		wait:   shouldWait,
+		reason: strings.Join(reasons, "; "),
+	}, nil
+}
+
+func (b Bidder) runResourceBidding(
+	ctx context.Context,
+	job *models.Job,
+	resources *models.Resources,
+) (*bidStrategyResponse, error) {
+	// calculate resource usage of the job, failure here represents a compute failure.
+	resourceUsage, err := b.usageCalculator.Calculate(ctx, *job, *resources)
 	if err != nil {
-		return nil, models.Resources{}, fmt.Errorf("error asking bidding strategy if we should bid: %w", err)
+		log.Ctx(ctx).Error().Err(err).Msg("Error calculating resource requirements for job")
+		return nil, fmt.Errorf("calculating resource usage of job: %w", err)
+	}
+
+	// ask the bidding strategy if we should bid on this job
+	request := bidstrategy.BidStrategyRequest{
+		NodeID:   b.nodeID,
+		Job:      *job,
+		Callback: b.getApproveURL(),
+	}
+
+	// assume we are bidding unless a request is rejected
+	shouldBid := true
+	// assume we're not waiting to bid unless a request indicates so.
+	shouldWait := false
+	reasons := make([]string, 0, len(b.resourceStrategy))
+
+	// TODO(forrest): this can be parallelized with a wait group, room for improvement here if resource validation
+	//  is time consuming.
```

Comment:
Should add a ticket so it can be measured and prioritised if it's relevant

Reply:
Same answer as here: https://github.com/bacalhau-project/bacalhau/pull/3680/files#r1543764567